Fix text selection from the panes

elvankonukseven opened the pull request:
This PR introduces tips/advice for users when they try selecting text from the panes. Every time a user makes the action of "selection" by clicking and scrolling on text from the panels, a temporary message displays.

The commands for the "selection" action is specific to terminals and this implementation checks for the terminal type to show the appropriate tip. 

Addresses issue #44

elvankonukseven committed:
fix: add advice(tip) for user when they try to select text, based on the terminal they use

elvankonukseven committed:
Remove unused import

elvankonukseven committed:
Remove unused function and variables from previous implementation tries

elvankonukseven committed:
Remove border

vgvassilev commented on src/idd/ui/scrollable_area.py:
```diff
@@ -10,19 +14,73 @@ class TextScrollView(RichLog):
     """
 
+    # Terminal-specific selection tips
+    SELECTION_TIPS = {
```
Can we detect which app we are running and display the relevant tip?

elvankonukseven replied:
Yes, this is what is done currently(?)
 For the "default" which is the Mac terminal, I haven't figured out the command and I don't think it's possible just with a command, that's why I put a general message for now.

vgvassilev replied:
Okay, sounds good. Do we need to enumerate the shortcuts by hand -- isn't there some api that we can use? I bet there is some way to configure these on the user end...

elvankonukseven replied:
I unfortunately couldn't find any api... I will definitely keep it in mind for later though

vgvassilev commented on src/idd/ui/scrollable_area.py:
```diff
@@ -10,19 +14,73 @@ class TextScrollView(RichLog):
         self.clear()
         self.append(lines)
+        
+    def _get_terminal_type(self):
```
Probably better to return some enum instead of comparing strings.

elvankonukseven replied:
os.environ returns a string so in the end the comparison will have to be between strings. Not sure if it's worth using enums ?

Vipul-Cariappa commented on src/idd/ui/scrollable_area.py:
```diff
@@ -10,19 +14,73 @@ class TextScrollView(RichLog):
+        "Terminal.app": "Option(⌥)+Click to select text",
+        "iTerm2": "Cmd(⌘)+Shift+C: Copy mode | Option(⌥)+Click: Selection",
+        "Warp": "Shift+Click to select text",
+        "default": "Use terminal's selection mechanism to copy text" #default mac terminal
```
```suggestion
        "default": "Use terminal's selection mechanism to copy text (Maybe Shift+Click)" #default mac terminal
```

Does this sound better?

Vipul-Cariappa commented on src/idd/ui/scrollable_area.py:
```diff
@@ -10,19 +14,73 @@ class TextScrollView(RichLog):
+    
+    def on_mouse_up(self, event: events.MouseUp) -> None:
+        """Reset click state when mouse button is released."""
+        if hasattr(self, '_clicked'):
```
Initialize `self._clicked` in `__init__`. And remove the `hasattr`.

elvankonukseven replied:
Thanks!

Vipul-Cariappa commented on src/idd/ui/scrollable_area.py:
```diff
@@ -10,19 +14,73 @@ class TextScrollView(RichLog):
+    def on_leave(self, event: events.Leave) -> None:
+        """Reset hover timer and click state when mouse leaves the widget."""
+        self._hover_start = 0
+        if hasattr(self, '_clicked'):
+            self._clicked = False
```
Same as above.

Vipul-Cariappa commented on src/idd/ui/scrollable_area.py:
```diff
@@ -10,19 +14,73 @@ class TextScrollView(RichLog):
+            
+            # Show tip after 0.1 seconds of hovering after a click
+            if hover_duration > 0.1:
+                terminal = self._get_terminal_type()
+                tip = self.SELECTION_TIPS.get(terminal, self.SELECTION_TIPS["default"])
```
Should be part of `__init__`. Why recompute this again?

elvankonukseven replied:
Makes sense, thanks!

Vipul-Cariappa commented on src/idd/ui/scrollable_area.py:
```diff
@@ -10,19 +14,73 @@ class TextScrollView(RichLog):
+    
+    def on_mouse_move(self, event: events.MouseMove) -> None:
+        """Show selection tip after clicking and hovering."""
+        # Only show tip if we've been clicked and haven't shown a tip yet
+        if hasattr(self, '_clicked') and self._clicked and not self.tip_shown:
```
Let's show the tip multiple times. I will need it, I can't remember anything 😄.

Vipul-Cariappa replied:
I did not realise you were modifying `TextScrollView`. I believe you will have to move all these changes to `DiffDebug`. It belongs there.

Vipul-Cariappa reviewed (commented):
Nice work. I have left some comments, please have a look.

vgvassilev commented:
I have even more basic question. Why standard copy and paste does not work? Why they are required to different shortcuts?

elvankonukseven commented:
@vgvassilev Indeed, it seems adding a tip message is the best way right now. The good news is I found how to do it in the Mac terminal. I'll add it too.

elvankonukseven committed:
Move logic into DiffDebug and initialize variables in init

elvankonukseven committed:
Remove implementation TextScrollView

elvankonukseven commented:
@vgvassilev  @Vipul-Cariappa The changes you suggested are implemented. I added the changes to DiffDebug and I added specific commands for vs code and Mac terminals. Vs code config was also about changing the user settings json with this: "terminal.integrated.macOptionClickForcesSelection": true. Not sure if I should add this in the tip ? Afterwards it just works with Opt+click.

Vipul-Cariappa reviewed (approved):
LGTM! Nice Work!

vgvassilev committed:
[ci] React on a tag push

elvankonukseven committed:
Merge branch 'main' into fix-copystrings
